README
Imlib2-Ruby 0.5.2 README
========================

This document was last updated on Sun May 29 18:34:43 2005.
Please see the file COPYING for licensing and warranty information.  The
latest version of this library is available at the following URL:
http://pablotron.org/software/imlib2-ruby/.

Introduction
============
Imlib2-Ruby is a Ruby interface to Imlib2.  All of Imlib2 (except for
X11 calls) has been wrapped.  I've provided a my test script and handful
of commented scripts in examples/, and there's an rdoc-generated API
reference in doc/ (which you can rebuild yourself by typing "rdoc
./imlib2.c").  Imlib2-Ruby is also available in the FreeBSD ports
collection (courtesy of Akinori Musha), as a NetBSD package (courtesy of 
Obata Akio), in Debian Linux via apt-get (courtesy of Jon Bernard), and
as a RubyGem (courtesy of Chad Fowler).  Imlib2-Ruby is still a work in
progress, so bug reports or patches are welcome.  Feel free to email me
at the address below if you have any additional questions.


System Requirements
===================
* Imlib2, version 1.2.0 (or newer):
  http://www.enlightenment.org/pages/imlib2.html
  (you can find source tarballs and RPMs at the following URL)
  ftp://ftp.sourceforge.net/pub/sourceforge/enlightenment/
* Ruby, version 1.8.x (or newer):
  http://www.ruby-lang.org/


Installation
============
ruby ./extconf.rb            # generate Makefile
make && su -c 'make install' # compile and install library


Related Links
=============
* RDoc: used to generate the API Documentation
  http://rdoc.sourceforge.net/
* RubyGems: Another method of installing Imlib2-Ruby
  http://rubygems.rubyforge.org/


Reporting Bugs
==============
Have a bug to report or a feature you'd like me to add to Imlib2-Ruby?
Feel free to email me at the address below.  Alternatively, you can
submit your feature request or bug directly to my bug-tracking web
interface at the following URL:

  http://bugs.pablotron.org/

(note that you'll need to create an account in order to submit a feature
request or a bug report via the web interface).


About the Author
================
Paul Duncan <pabs@pablotron.org>
http://pablotron.org/

Please see the file AUTHORS for additional contributions.


ChangeLog
* Wed Mar 27 11:04:01 2002, pabs <pabs@pablotron.org>
  * released version 0.1.0 yesterday

* Thu Mar 28 20:20:11 2002, pabs <pabs@pablotron.org>
  * added ellipse routines (Imlib2::Image#{draw,fill}_ellipse)
  * added rotation routines (Imlib2::Image#rotate{,!})
  * updated checkerboard.rb to draw pieces as well (to test
    {draw,fill}_ellipse)

* Sat Mar 30 15:07:02 2002, pabs <pabs@pablotron.org>
  * added draw_oval and fill_oval (more concice aliases for *_ellipse)
  * better comments in examples/checkerboard.rb
  * fixed blue/green mixup in set_context_color()
  * added object-oriented version of examples/checkerboard.rb
    (examples/checkerboard2.rb)
  * updated doc/api.txt with new ellipse calls

* Sat Mar 30 19:11:56 2002 , pabs <pabs@pablotron.org>
  * added Imlib2::Font (and associated methods)
  * added Imlib2::Cache  (and associated methods)
  * added Imlib2::Font::{add,remove}_path, Imlib2::Font::paths

* Sat Mar 30 20:11:18 2002, pabs <pabs@pablotron.org>
  * added Imlib2::Image#draw_text() -- works as follows:
    im.draw_text font, text, [x, y]
    im.draw_text font, text, [x, y], color
    im.draw_text font, text, {'x' => x, 'y' => y}
    im.draw_text font, text, {'x' => x, 'y' => y}, color
    im.draw_text font, text, x, y
    im.draw_text font, text, x, y, color
  * fixed color setting routines (had 0 for argument count)

* Sat Mar 30 21:04:46 2002, pabs <pabs@pablotron.org>
  * added examples/test_font.rb
  * added font, cache, and font path calls to doc/api.txt

* Sat Mar 30 22:08:32 2002, pabs <pabs@pablotron.org>
  * added better comments in examples/test_font.rb
  * released version 0.2.0

* Sun Mar 31 06:04:06 2002, pabs <pabs@pablotron.org>
  * added Imlib2::Gradient::* (Imlib_Color_Range)
  * added Imlib2::Image#{,fill_}gradient
  * examples/checkerboard.rb now uses a Gradient for the foreground color
  * added Imlib2::ColorRange (inherits from Imlib2::Gradient, so it's
    essentially an alias for Gradient)

* Sun Mar 31 06:46:36 2002, pabs <pabs@pablotron.org>
  * added Imlib2::Polygon::*
  * added examples/test_poly.rb
  * renamed examples/checkerboard2.rb to examples/oo-board.rb

* Sun Mar 31 07:50:19 2002, pabs <pabs@pablotron.org>
  * added Imlib2::Image::blend_image! and Imlib2::Image::blend_image
    (and also their more concise synonyms: blend! and blend, respectively)

* Sun Mar 31 16:42:32 2002, pabs <pabs@pablotron.org>
  * released 0.3.0

* Sun Apr 14 02:17:11 2002, pabs <pabs@pablotron.org>
  * inline documentation (for rdoc) finished for Imlib2::Border,
    Imlib2::Cache, and Imlib2::Color::RgbaColor.

* Sun Apr 14 04:01:55 2002, pabs <pabs@pablotron.org>
  * inline documentation for Imlib2::Color::HsvaColor
  * changed border.{l,l=,r,r=,t,t=,b,b=} to aliases instead of method
    defines

* Wed Apr 17 15:58:40 2002, pabs <pabs@pablotron.org>
  * added Imlib2::Filter and associated methods
  * Imlib2::Image#filter (it's polymorphic; takes either a String,
    which it passes to imlib_apply_filter(), or an Imlib2::Filter,
    which it passes to imlib_image_filter()).  the explicit calls 
    are Imlib2::Image#string_filter and Imlib2::Image#static_filter,
    respectively.

* Wed Apr 17 16:35:36 2002, pabs <pabs@pablotron.org>
  * added support for text directions (via Imlib2::Direction::LEFT,
    Imlib2::Direction::RIGHT, etc).  it's now an optional last argument
    to Imlib2::Image::#draw_text

* Thu Apr 18 14:06:17 2002, pabs <pabs@pablotron.org>
  * added Imlib2::ColorModifier
  * added Imlib2::Image#apply_color_modifier and Imlib2::Image#apply_cmod
    they take either a color modifier or a color modifier and a rectangle
  * added Imlib2::Encoding (font encoding constants)
  * initial Imlib2::Context code.

* Thu Apr 18 15:44:57 2002, pabs <pabs@pablotron.org>
  * all non-X11 context calls available, except for progress callbacks.

* Thu Apr 18 19:34:38 2002, pabs <pabs@pablotron.org>
  * updated README to reflect post-0.2.0 features.
  * applied extconf.rb patch from Akinori Musha <knu@iDaemons.org>
  * added AUTHORS.

* Thu Apr 18 20:09:59 2002, pabs <pabs@pablotron.org>
  * documentation for Imlib2::Color::HlsaColor methods
  * documentation for Imlib2::Color::CmyaColor methods
  * documentation for Imlib2::Image::load_* methods

* Fri Apr 19 00:05:04 2002, pabs <pabs@pablotron.org>
  * documentation for all Imlib2::Image methods
  * documentation for all Imlib2::Font methods
  * documentation for all Imlib2::Gradient methods
  * documentation for all Imlib2::Filter methods
  * documentation for all Imlib2::Context methods

* Fri Apr 19 01:22:06 2002, pabs <pabs@pablotron.org>
  * updated TODO
  * posted docs online (http://www.pablotron.org/software/imlib2-ruby/doc/)
  * updated README to mention doc/ dir
  * copied ChangeLog online
  * released 0.4.0

* Mon Apr 22 11:48:33 2002, pabs <pabs@pablotron.org>
  * fixed a bug in draw_pixel reported by Phlip (phlip_cpp@yahoo.com).
  * updated Imlib2 requirement in README, and added a link to the source
    download (via Sourceforge FTP).
  * change Imlib2::Context get methods so they don't require an argument
    (oops :D)
  * added workaround for broken imlib_image_draw_pixel() call (uses
    imlib_image_draw_rectangle() instead)
  * released 0.4.1

* Wed Apr 24 22:00:37 2002, pabs <pabs@pablotron.org>
  * removed tutorial/ directory (how the hell did that get in the main
    release?)
  * minor fix to examples/checkerboard.rb (and examples/oo-board.rb) to
    avoid imlib2 fill_ellipse bug

* Sun Apr 28 21:48:36 2002, pabs <pabs@pablotron.org>
  * added Imlib2::Image::draw_pixel_workaround{?,=} to enable or disable
    imlib_image_draw_pixel() work-around behavior
  * changed minor number to 5 (0.5.0)
  * added X11 calls
  * partial support for X11 primitives in Imlib2::X11, although I doubt
    they interact properly with Ruby/X11 or Ruby/Xlib properly.
  * TODO: Ruby/Gtk+ support (via gdkx primitive wrappers)
  * changed all instances (except one, where it was appropriate) of

examples/test_font.rb
#!/usr/bin/ruby

########################################################################
# test_font.rb - draw translucent text on an image                     #
# (c) 2002 Paul Duncan <pabs@pablotron.org>                            #
#                                                                      #
# - test Imlib2::Font methods and Imlib2::Color methods                #
########################################################################

# load imlib2 binding
require 'imlib2'

# text to draw
text = 'Bill Gates is a criminal!'

# filename variables
filename = 'images/bill_gates.jpg'
output_filename = 'test_font_1.png'
output_filename = ARGV.shift if ARGV.length != 0

# font variables
fontpath = 'fonts'
# these fonts were removed due to licensing issues
#fontname = 'quadapto/32'
#fontname = '20thcent/32'
fontname = 'yudit/32'

# font colors
colors = [ 
  Imlib2::Color::RED,
  Imlib2::Color::YELLOW,
  Imlib2::Color::GREEN,
  Imlib2::Color::AQUA,
  Imlib2::Color::BLUE,
  Imlib2::Color::VIOLET,
].each { |color| color.a = 128 }

# text drop-shadow x offset, y offset, and color
sh = {
  'x'     => 2,
  'y'     => 2,
  'color' => Imlib2::Color::RgbaColor.new(0, 0, 0, 64),
}

# load filename
begin
  im = Imlib2::Image::load filename
rescue Imlib2::FileError
  die "Couldn't load \"#{filename}\": " + $!
end

# add specified font path and load font
Imlib2::Font::add_path fontpath
font = Imlib2::Font.new fontname

# get the height and width of the specified text with our font, and use that
# to calculate the x and y offset so the text will be centered 
fw, fh = font.size text
fx, fy = (im.width - fw) / 2, (im.height / 2 - fh) / 2;

# draw text on image with specified font and color
c_i = -1
0.upto(im.height / fh) { |offset|
  # draw drop-shadow
  im.draw_text font, text, fx + sh['x'], fh * offset + sh['y'], sh['color']

  # draw text (in specified color)
  color = colors[c_i = (c_i + 1) % colors.length]
  im.draw_text font, text, fx, fh * offset, color
}

# save image to output filename
begin
  im.save output_filename
rescue Imlib2::FileError
  $stderr.puts "Couldn't save \"#{output_filename}\": " + $!
end


examples/test_poly.rb
#!/usr/bin/ruby

require 'imlib2'

filename = 'test_poly_1.png'

im = Imlib2::Image.new 320, 320
fg = Imlib2::Color::BLUE
bg = Imlib2::Gradient.new [0, Imlib2::Color::BLACK],
                          [1, Imlib2::Color::WHITE]

im.gradient bg, 0, 0, im.width, im.height, 135.0


points = []

0.upto(im.height / 10) { |y|
  points.push [5, y * im.height / 10], [im.width - 5, y * im.height / 10] 
}

poly = Imlib2::Polygon::new *points

im.fill_poly poly, fg

begin
  im.save filename
rescue Imlib2::FileError
  $stderr.puts "Couldn't save file \"#{filename}\": " + $!
end


examples/test_attach.rb
#!/usr/bin/env ruby

OUT_PATH = 'attach.jpg'

require 'imlib2'

# load image
im_path = ARGV.shift || (File::dirname($0) << '/images/bill_gates.jpg')
puts "Loading input image \"#{im_path}\"..."
im = Imlib2::Image.load(im_path)

# assign key, generate random number
key, val = (ARGV.shift || 'asdf'), (ARGV.shift || rand(10000)).to_i

puts "random number: #{val}"
puts "key: #{key}"
im[key] = val

puts "testing get_attach_value: #{im[key]}"

# save output image
puts 'Saving output image...'
im.save(OUT_PATH)


extconf.rb
require 'mkmf'

imlib2_config = with_config("imlib2-config", "imlib2-config")

$CFLAGS << ' -DX_DISPLAY_MISSING ' << `#{imlib2_config} --cflags`.chomp
$LDFLAGS << ' ' << `#{imlib2_config} --libs`.chomp

$LDFLAGS.gsub!(/\ -lX11\ -lXext/, "") if RUBY_PLATFORM =~ /darwin/

if have_library("Imlib2", "imlib_create_image")
  # test for faulty versions of imlib2
  ver = `imlib2-config --version`.chomp.split(/\./)
  major, minor, revision = ver[0].to_i, ver[1].to_i, ver[2].to_i
  if ((major > 1)               ||
      (major == 1 && minor > 0) ||
      (major == 1 && minor == 0 && revision > 5))
    $CFLAGS << ' -DDISABLE_DRAW_PIXEL_WORKAROUND '
  else
    puts 'Note: This version of Imlib2 has a bug in imlib_image_draw_pixel().',
         'Enabling workaround (see documentation for details).'
  end

  create_makefile("imlib2")
end


examples/oo-board.rb
#!/usr/bin/ruby

########################################################################
# oo-board.rb - draw a checkerboard (object-oriented version)          #
# (c) 2002 Paul Duncan <pabs@pablotron.org>                            #
#                                                                      #
# - test Imlib2::Image#fill_rect, Imlib2::Image#save, and color        #
#   constants (Imlib2::Color::BLACK, Imlib2::Color::WHITE, etc)        #
########################################################################

# load imlib2 bindings
require "imlib2"

class CheckerBoard
  attr_accessor :w, :h, :bw, :bh, :fg, :bg, :tc, :bc

  def initialize(*args)
    # number, width, and height of boxes
    @w, @h, @bw, @bh = args
    
    # foreground, background, top piece, and bottom piece colors
    @fg, @bg = Imlib2::Color::WHITE, Imlib2::Color::BLACK
    @tc, @bc = Imlib2::Color::RED, Imlib2::Color::BLUE
    
    # allocate the new image and fill it with the background color
    @im = Imlib2::Image.new @w * @bw, @h * @bh
    @im.fill_rect [0, 0], [@w * @bw, @h * @bh], @bg
  end

  # fill squares with foreground color
  def draw_squares()
    0.upto(@w / 2) { |x|
      0.upto(@h / 2) { |y|
        (0..1).each { |i|
          rect = [x * @bw * 2 + @bw * i, y * @bh * 2 + @bh * i, @bw, @bh]
          @im.fill_rect rect, @fg
        }
      }
    }
  end

  # draw checker pieces
  def draw_pieces()
    0.upto(@w - 1) { |x|
      0.upto(2) { |y|
        # draw top piece
        rect = [x * @bw + @bw / 2, y * @bh + @bh / 2, @bw * 0.4, @bh * 0.4]
        # Imlib2 bug (works w/ draw_ellipse, but not w/ fill_ellipse)
        @im.fill_ellipse rect, @tc if (x - (y % 2)) % 2 == 1
    
        # draw bottom piece
        rect[1] = h * @bh - rect[1]
        # Imlib2 bug (works w/ draw_ellipse, but not w/ fill_ellipse)
        @im.fill_ellipse rect, @bc if (x - (y % 2)) % 2 == 0
      }
    }
  end

  # save image
  def save(filename)
    @im.save filename
  end
end

# if this file is the executable, then test out our member functions by
# creating a CheckerBoard instance and saving it as 'checkerboard.png'
# (unless a filename was passed on the command line)
if $0 == __FILE__
  # set output filename (use argument if specified)
  filename = "checkerboard.png"
  filename = ARGV.pop if ARGV.length > 0

  board = CheckerBoard.new 10, 10, 48, 48
  
  puts "Generating checkerboard..."
  board.draw_squares

  puts "Generating pieces..."
  board.draw_pieces
  
  # save image (with error checking)
  puts "Saving image to \"#{filename}\"..."
  begin
    board.save filename
  rescue Imlib2::FileError
    $stderr.puts "Couldn't save \"#{filename}\": " + $!
  end

  puts "Done."
end


examples/checkerboard.rb
#!/usr/bin/ruby

########################################################################
# checkerboard.rb - draw a checkerboard                                #
# (c) 2002 Paul Duncan <pabs@pablotron.org>                            #
#                                                                      #
# - test Imlib2::Image#fill_rect, Imlib2::Image#save, and color        #
#   constants (Imlib2::Color::BLACK, Imlib2::Color::WHITE, etc)        #
########################################################################

# load imlib2 bindings
require "imlib2"

# number, width, and height of boxes
w, h, bw, bh = 10, 10, 48, 48

# foreground (gradient), background, top piece, and bottom piece colors
fg = [Imlib2::Color::DARKGREY, Imlib2::Color::LIGHTGREY]
bg = Imlib2::Color::BLACK
tc, bc = Imlib2::Color::RED, Imlib2::Color::BLUE

# set output filename (use command-line argument if specified)
filename = "checkerboard.png"
filename = ARGV.pop if ARGV.length > 0

# allocate the new image and fill it with the background color
im = Imlib2::Image.new w * bw, h * bh
im.fill_rect [0, 0], [w * bw, h * bh], bg

# create foreground gradient
gradient = Imlib2::Gradient.new [0, fg[0]], [1, fg[1]]

# fill squares with foreground gradient
puts "Generating checkerboard..."
0.upto(w / 2) { |x|
  0.upto(h / 2) { |y|
    (0..1).each { |i|
      rect = [x * bw * 2 + bw * i, y * bh * 2 + bh * i, bw, bh]
      im.fill_gradient gradient, rect, 135.0
    }
  }
}

# draw checker pieces
puts "Generating pieces..."
0.upto(w - 1) { |x|
  0.upto(2) { |y|
    # draw top piece
    rect = [x * bw + bw/2, y * bh + bh/2, bw * 2/5, bh * 2/5]
    # Imlib2 bug (works properly w/ draw_ellipse, but not w/ fill_ellipse)
    im.fill_ellipse rect, tc if (x - (y % 2)) % 2 == 1

    # draw bottom piece
    rect[1] = h * bh - rect[1]
    # Imlib2 bug (works properly w/ draw_ellipse, but not w/ fill_ellipse)
    im.fill_ellipse rect, bc if (x - (y % 2)) % 2 == 0
  }
}

# save image (with error checking)
puts "Saving to \"#{filename}\"."
begin
  im.save filename
rescue Imlib2::FileError
  $stderr.puts "Couldn't save \"#{filename}\": " + $!
end
